core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -615,31 +640,75 @@ public void close() {
     }
 
     static final class PoolKey {
-        final String host;
-        @Nullable
-        final String ipAddr;
-        final int port;
-        final int hashCode;
+        final Endpoint endpoint;
         final ProxyConfig proxyConfig;
+        final SocketAddress remoteAddress;
+
+        private final int hashCode;
+        private final String strVal;
 
-        PoolKey(String host, @Nullable String ipAddr, int port, ProxyConfig proxyConfig) {
-            this.host = host;
-            this.ipAddr = ipAddr;
-            this.port = port;
+        PoolKey(Endpoint endpoint, ProxyConfig proxyConfig) {
+            this.endpoint = endpoint;
             this.proxyConfig = proxyConfig;
-            hashCode = Objects.hash(host, ipAddr, port, proxyConfig);
+            remoteAddress = toRemoteAddress(endpoint, proxyConfig);
+            hashCode = endpoint.hashCode() * 31 + proxyConfig.hashCode();
+            strVal = generateString(endpoint, proxyConfig);
+        }
+
+        private static String generateString(Endpoint endpoint, ProxyConfig proxyConfig) {
+            final String host = endpoint.host();
+            final String ipAddr = endpoint.ipAddr();
+            final int port = endpoint.port();
+            final boolean isDomainSocket = endpoint.isDomainSocket();
+            final String proxyConfigStr = proxyConfig.proxyType() != ProxyType.DIRECT ? proxyConfig.toString()
+                                                                                      : null;
+            try (TemporaryThreadLocals tempThreadLocals = TemporaryThreadLocals.acquire()) {
+                final StringBuilder buf = tempThreadLocals.stringBuilder();
+                buf.append('{').append(host);
+                if (ipAddr != null) {
+                    buf.append('/').append(ipAddr);
+                }
+                if (!isDomainSocket) {
+                    buf.append(':').append(port);
+                }
+                if (proxyConfigStr != null) {
+                    buf.append(" via ");
+                    buf.append(proxyConfigStr);
+                }
+                buf.append('}');
+                return buf.toString();
+            }
         }
 
-        private InetSocketAddress toRemoteAddress() throws UnknownHostException {
+        private static SocketAddress toRemoteAddress(Endpoint endpoint, ProxyConfig proxyConfig) {
```

Comment:
Most parts of this method could be replaced with the brand new `Endpoint.toSocketAddress(0)`.